packages/docs-reanimated/docs/guides/worklets.mdx

```diff
@@ -0,0 +1,142 @@
+
+Most of the time when working with Reanimated and [Gesture Handler](https://docs.swmansion.com/react-native-gesture-handler/) the code is automatically workletized and run on the UI thread by default.
+
+```javascript
+import { useAnimatedStyle } from 'react-native-reanimated';
+
+function App() {
+  const style = useAnimatedStyle(() => {
+    // Running on the UI thread
+    return { opacity: 0.5 };
+  });
+}
+```
+
+The `"worklet";` directive has no effect on the [JavaScript thread](/docs/fundamentals/glossary#javascript-thread). Worklets on the JavaScript thread are just functions.
+
+You can use [`runOnUI`](/docs/threading/runOnUI) to manually run worklets on the UI thread:
+
+```javascript
+function myWorklet() {
+  'worklet';
+  console.log('Hello from the UI thread');
+}
+
+function onPress() {
+  runOnUI(myWorklet)();
+}
+```
+
+You can pass data as arguments to worklets.
+
+```javascript
+function myWorklet(greeting) {
+  'worklet';
+  console.log(`${greeting} from the UI thread`);
+}
+
+function onPress() {
+  runOnUI(myWorklet)('Howdy');
+}
+```
+
+Worklets are [closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures). They can access variables declared outside of it's own scope. Only variables referenced in the worklet body will be captured inside the worklet scope.
+
+```javascript
+const width = 135.5;
+
+function otherWorklet() {
+  'worklet';
+  console.log('Captured width is', width);
+}
+```
+
+Capturing big JavaScript objects inside of a worklet can lead to performance issues.
+
+```javascript
+const theme = {...}; // theme is a big object
+
+function myWorklet() {
+  'worklet';
+  console.log(theme.color); // 🚨 referenced `color` but captured the whole `theme` object
+}
+```
+
+You can get around this problem by first assigning the prop you want to use to a separate variable .
+
+```javascript
+const theme = {...};
+const color = theme.color;
+
+function myWorklet() {
+  'worklet';
+  console.log(color); // ✅ captured only `color`
+}
+```
+
+Worklets can return data within the same thread.
+
+```javascript
+function returningWorklet() {
+  'worklet';
+  return "I'm back"; // on the UI thread
+}
+
+function someWorklet() {
+  'worklet';
+  let what = returningWorklet(); // still on the UI thread
+  console.log('On the UI thread, other worklet says', what);
+}
+```
+
+To pass data between UI and JS thread we use [shared values](/docs/fundamentals/glossary#shared-value).
+
+```javascript
+import { useSharedValue } from 'react-native-reanimated';
+
+function App() {
+  const width = useSharedValue(100);
+
+  function myWorklet() {
+    'worklet';
+    width.value += 50;
+  }
+
+  width.value; // available on both JS and UI thread
```

Comment:
Actually, you should never read or write shared value during a render.

Reply:
+1, let's put it in a `React.useEffect` callback

Reply:
+2